src/graph.cc

```diff
@@ -163,12 +163,28 @@ bool DependencyScan::RecomputeOutputDirty(Edge* edge,
     }
 
     if (output_mtime < most_recent_input->mtime()) {
-      EXPLAIN("%soutput %s older than most recent input %s "
-              "(%d vs %d)",
-              used_restat ? "restat of " : "", output->path().c_str(),
-              most_recent_input->path().c_str(),
-              output_mtime, most_recent_input->mtime());
-      return true;
+      bool really_dirty = true;
+      if (edge->GetBindingBool("hash_input")) {
+        string err;
+        really_dirty = hash_log().EdgeChanged(edge, &err);
+        if (!err.empty()) {
+          really_dirty = true; // in case of an error we behave as if hashing
+          Error(err.c_str());  // would be disabled, but we log the error
+        }
+      }
+
+      if (really_dirty) {
+        EXPLAIN("%soutput %s older than most recent input %s "
+                "(%d vs %d)",
+                used_restat ? "restat of " : "", output->path().c_str(),
+                most_recent_input->path().c_str(),
+                output_mtime, most_recent_input->mtime());
+        return true;
+      } else {
+        EXPLAIN("output %s older than most recent input, but hashed "
+                "contents of all inputs are unchanged",
+              output->path().c_str());
+      }
```

Comment:
the 'return true' statements are early exits in case the edge is already identified dirty.
if not yet determined we have to fall through to the next criteria. in this case the next criteria is the command line (changed/unchanged). if we would exit with false here, any subsequent check would be ignored.

please note the function end is further down in line 192/208 and not in line 173/189